tests/kafkatest/tests/streams/streams_smoke_test.py

```diff
@@ -31,48 +31,80 @@ def __init__(self, test_context):
         })
 
+        self.test_context = test_context
         self.driver = StreamsSmokeTestDriverService(test_context, self.kafka)
-        self.processor1 = StreamsSmokeTestJobRunnerService(test_context, self.kafka)
-        self.processor2 = StreamsSmokeTestJobRunnerService(test_context, self.kafka)
-        self.processor3 = StreamsSmokeTestJobRunnerService(test_context, self.kafka)
-        self.processor4 = StreamsSmokeTestJobRunnerService(test_context, self.kafka)
 
-    @cluster(num_nodes=9)
-    def test_streams(self):
-        """
-        Start a few smoke test clients, then repeat start a new one, stop (cleanly) running one a few times.
-        Ensure that all results (stats on values computed by Kafka Streams) are correct.
-        """
+    @cluster(num_nodes=8)
+    @matrix(eos=[True, False], crash=[True, False])
+    def test_streams(self, eos, crash):
+        #
+        if eos:
+            processor1 = StreamsSmokeTestEOSJobRunnerService(self.test_context, self.kafka)
+            processor2 = StreamsSmokeTestEOSJobRunnerService(self.test_context, self.kafka)
+            processor3 = StreamsSmokeTestEOSJobRunnerService(self.test_context, self.kafka)
+        else:
+            processor1 = StreamsSmokeTestJobRunnerService(self.test_context, self.kafka)
+            processor2 = StreamsSmokeTestJobRunnerService(self.test_context, self.kafka)
+            processor3 = StreamsSmokeTestJobRunnerService(self.test_context, self.kafka)
 
-        self.driver.start()
 
-        self.processor1.start()
-        self.processor2.start()
 
-        time.sleep(15)
+        with processor1.node.account.monitor_log(processor1.STDOUT_FILE) as monitor1:
+            processor1.start()
+            monitor1.wait_until('REBALANCING -> RUNNING',
+                               timeout_sec=60,
+                               err_msg="Never saw 'REBALANCING -> RUNNING' message " + str(processor1.node.account)
+                               )
 
-        self.processor3.start()
-        self.processor1.stop()
+            self.driver.start()
 
-        time.sleep(15)
+            monitor1.wait_until('processed',
+                                timeout_sec=30,
+                                err_msg="Didn't see any processing messages " + str(processor1.node.account)
+                                )
 
-        self.processor4.start()
+            # make sure we're not already done processing (which would invalidate the test)
+            self.driver.node.account.ssh("! grep 'Result Verification' %s" % self.driver.STDOUT_FILE, allow_fail=False)
+
+            processor1.stop_nodes(not crash)
+
+        with processor2.node.account.monitor_log(processor2.STDOUT_FILE) as monitor3:
+            processor2.start()
+            monitor3.wait_until('REBALANCING -> RUNNING',
+                                timeout_sec=120,
+                                err_msg="Never saw 'REBALANCING -> RUNNING' message " + str(processor2.node.account)
```

Comment:
Again, wait for the node to start. This isn't strictly necessary, since we're just going to wait for "processed" next, but I thought it made the tests more readable and debuggable. For example, when you get a test failure, you'll know whether the node failed to start or whether it started, but didn't process anything.

Reply:
Not a comment: I actually think it is indeed necessary to avoid flakiness, remember @bbejeck once talked about it in an older PR?

Reply:
I think that was a different situation, like maybe we stopped and then re-started a node? In that case, you need to grab a new monitor after you stop and before you re-start, to be sure your grep won't match the "started" message from the first time it was running.

Since we're querying the same monitor here in both blocks, we're grepping over the same range of the file. The monitor pins the start-point for grepping when you create it. For example, since we create this monitor before we start Streams, it uses "byte 1" as the starting points for all greps for both `REBALANCING -> RUNNING` and `processed`.